src/thumbnailer_slope_optim.cc

```
@@ -0,0 +1,253 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "thumbnailer.h"
+
+namespace libwebp {
+
+Thumbnailer::Status Thumbnailer::GenerateAnimationSlopeOptim(
+    WebPData* const webp_data) {
+  Thumbnailer::Status anim_status = LossyEncodeSlopeOptim(webp_data);
+  if (anim_status == kOk) {
+    anim_status = TryNearLossless(webp_data);
+  }
+
+  if (anim_status == kOk) {
+    anim_status = LossyEncodeNoSlopeOptim(webp_data);
+  }
+
+  return anim_status;
+}
+
+Thumbnailer::Status Thumbnailer::FindMedianSlope(float* const median_slope) {
+  std::vector<float> slopes;
+
+  for (auto& frame : frames_) {
+    frame.config.quality = 100;
+    float psnr_100;  // pic's psnr value with quality = 100.
+    int size_100;    // pic'size with quality = 100.
+    if (GetPictureStats(frame.pic, frame.config, &size_100, &psnr_100) != kOk) {
+      return kStatsError;
```

I think in this kind of situation you should propagate the original error you got instead of returning a new one. Otherwise the original error is lost. It might be useful to have a macro as I've mentioned in a previous review (it was a mistake from me at the time since the error types were different, but here the error types are the same).
Something like:
#define CHECK_THUMBNAILER_STATUS(status)              \
  do {                                        \
    const Thumbnailer::Status  __S__ = (status);         \
    if (__S__ != Thumbnailer::kOk) {             \
      return __S__;                           \
    }                                         \
  } while (0)

Then you can just do:
CHECK_THUMBNAILER_STATUS(GetPictureStats(frame.pic, frame.config, &size_100, &psnr_100));